### PR TITLE
Submit Parallel Launcher to FlatHub

### PR DESCRIPTION
**Pull request by mpharoah**
Submitting [Parallel Launcher](https://parallel-launcher.ca/) to FlatHub. This application depends on RetroArch, so it bundles it as well.

### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [inline ca.parallel_launcher.ParallelLauncher.json:39]: All permissions here are the same as the ones required by [RetroArch](https://github.com/flathub/org.libretro.RetroArch) (which this application launches). One exception is that I removed the `--filesystem=xdg-run/app/com.discordapp.Discord:create` permission, because I don't need RetroArch to have that for my use case.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [inline ca.parallel_launcher.ParallelLauncher.json:38]: finish-args is usually above "modules" in the manifest, as a common usage.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [inline ca.parallel_launcher.ParallelLauncher.json:46]: this is really broad filesystem permission, and redundant.

File: flathub.json
Comment (hfiguiere) [inline flathub.json:2]: this is really kind unfortunate.

File: retroarch/retroarch.json
Comment (hfiguiere) [inline retroarch/retroarch.json:31]: this is in the runtime

File: retroarch/retroarch.json
Comment (hfiguiere) [inline retroarch/retroarch.json:33]: you should use the shared module

File: retroarch/retroarch.json
Comment (hfiguiere) [inline retroarch/retroarch.json:21]: what's wrong with the libpng in the runtime?

File: retroarch/retroarch.json
Comment (hfiguiere) [inline retroarch/retroarch.json:27]: 2.0,15 is in the runtime. why rebuilding it.

**Review by hfiguiere — CHANGES_REQUESTED** [review]
I don't understand why you'd rebuild RetroArch when it's available as a  flatpak.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [inline ca.parallel_launcher.ParallelLauncher.json:36]: there is no need to have these as indiviual manifest files. it makes maintenance harder.

File: retroarch/retroarch.json
Comment (hfiguiere) [inline retroarch/retroarch.json:9]:
```suggestion
			"commit": "9b282aa742b6c3d2f2925ae5a12e2cd7c6b6ad38"
```

commit ain't a branch

**Comment by hfiguiere** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

**Comment by mpharoah** [comment]
> I don't understand why you'd rebuild RetroArch when it's available as a flatpak.
Is it possible to specify another flatpak as a dependency, and then invoke it from another application within the Flatpak sandbox?

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:46]: All of these permissions are from the RetroArch flatpak, which has already been approved. It requires them because it launches RetroArch

File: retroarch/retroarch.json
Comment (mpharoah) [reply retroarch/retroarch.json:33]: It is. These are all from the [RetroArch flatpak](https://github.com/flathub/org.libretro.RetroArch/blob/master/org.libretro.RetroArch.json#L57)

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [reply ca.parallel_launcher.ParallelLauncher.json:46]: after approval a lot of things happen without.

Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:46]: This one is needed by RetroArch. It's also needed by Parallel Launcher itself because you can add any directory on your machine as a location to search for roms (optionally recursively).

File: retroarch/retroarch.json
Comment (mpharoah) [reply retroarch/retroarch.json:33]: It looks like the initial PR for adding RetroArch didn't have any of these. I wonder if they are actually required. They might be needed for bugfixes, but I'm not sure. I'll see if I can build without them and test it to make sure it works as expected.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:36]: Is it possible to include more than one module in a module file (other than the root one)?

**Comment by mpharoah** [comment]
Any idea why the test build is failing? The submodule its trying to clone is public-- in fact it's in this very org.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Fixed

File: retroarch/retroarch.json
Comment (mpharoah) [reply retroarch/retroarch.json]: Fixed

**Comment by mpharoah** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

File: retroarch/retroarch.json
Comment (mpharoah) [reply retroarch/retroarch.json:33]: Okay, it looks like the flatpak build wasn't actually even building any of these, so they clearly aren't required. I don't know why the RetroArch maintainers added them. Everything seems to be working fine without them, so I will remove them.

File: retroarch/retroarch.json
Comment (mpharoah) [reply retroarch/retroarch.json:21]: As it turns out, nothing. I just copied this from what the RetroArch flatpak does, but it seems to be unnecessary. Will remove.

**Comment by mpharoah** [comment]
When launching RetroArch from Parallel Launcher in my local flatpak installation, I get this warning printed to the console:
> sh: line 1: xdg-mime: command not found

It appears to be harmless as everything functions correctly, but is this something I should add? Or is there a reason it was excluded from the runtime?

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Oh, I see `home` is redundant with `host`. I will fix that.

Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Fixed

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Nevermind, added this permission back since I do actually need it

**Comment by mpharoah** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [inline ca.parallel_launcher.ParallelLauncher.json:15]: why do you need this?

File: flathub.json
Comment (hfiguiere) [reply flathub.json:2]: so why are you limiting to x86_64?

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:15]: RetroArch uses it, and I launch RetroArch

**Comment by mpharoah** [comment]
I have only tested it on `x86_64` and currenly only provide the custom emulator core on 64-bit for Linux. I'm not comfortable releasing it for all architectures yet since I haven't tested them. I will add more as I test on more hardware.

File: flathub.json
Comment (mpharoah) [reply flathub.json:2]: I have only tested it on `x86_64` and currenly only provide the custom emulator core on 64-bit for Linux. I'm not comfortable releasing it for all architectures yet since I haven't tested them. I will add more as I test on more hardware. All of my current users on Linux use x86_64.

Comment (hfiguiere) [reply flathub.json:2]: aarch64 is 64-bits linux too.

**Comment by hfiguiere** [comment]
I am still skeptical of why you need to build retroarch in this flatpak as it is already available on flathub. But maybe I misunderstand what the launcher does.

File: flathub.json
Comment (hfiguiere) [reply flathub.json:2]: and retroarch is available on aarch64.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [inline ca.parallel_launcher.ParallelLauncher.json:80]: this is the exact same source tree as the previous two modules.

File: flathub.json
Comment (mpharoah) [reply flathub.json:2]: Yes, but I haven't tested it to verify that it works as expected, so I won't release it for that arch until I can verify I have done endianness detection correctly everywhere and aren't accidentally relying on x86 somewhere.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:80]: I believe that's to extract the subdirectory. The RetroArch flatpak does the same thing.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (hfiguiere) [reply ca.parallel_launcher.ParallelLauncher.json:15]: it still doesn't explain why.

**Comment by mpharoah** [comment]
> I am still skeptical of why you need to build retroarch in this flatpak as it is already available on flathub
AFAIK there is no way for a flatpak to have a dependency on another flatpak being installed. Bundling it here also has the advantage that I can ensure the version of RetroArch being launched is one that I have tested and I know works.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:80]: Actually, it might be a goof by the RetroArch maintainers. I'll trying removing them and seeing if everything still works.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json:15]: Hmm, since I only ever run the N64 emulator cores, which are 64-bit, I may be able to remove this permission.

Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Yep, it isn't required for my use case. Removed it.

File: ca.parallel_launcher.ParallelLauncher.json
Comment (mpharoah) [reply ca.parallel_launcher.ParallelLauncher.json]: Yeah, turns out it isn't needed at all.

**Comment by mpharoah** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

**Comment by mpharoah** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

**Comment by mpharoah** [comment]
bot, build ca.parallel_launcher.ParallelLauncher

**Pull request closed without merging**
